python/graphstorm/gconstruct/utils.py

```diff
@@ -30,6 +30,110 @@
 from ..utils import sys_tracker
 from .file_io import HDF5Array
 
+SHARED_MEM_OBJECT_THRESHOLD = 1.9 * 1024 * 1024 * 1024 # must < 2GB
+SHARED_MEMORY_CROSS_PROCESS_STORAGE = "shared_memory"
+PICKLE_CROSS_PROCESS_STORAGE = "pickle"
+
+def _to_shared_memory(data):
+    """ Move all tensor objects into torch shared memory
+
+    Parameters
+    ----------
+    data: dict/tuple/list of tensors
+        Data returned by user_parser
+    """
+    if th.is_tensor(data):
+        return data.share_memory_()
+    elif isinstance(data, np.ndarray):
+        # only handle data that can be converted to torch tensor
+        if data.dtype in [np.float64, np.float32, np.float16,
+                          np.complex64, np.complex128, np.int64,
+                          np.int32, np.int16, np.int8,
+                          np.uint8]:
+            return th.tensor(data).share_memory_()
+        return data
+    elif isinstance(data, dict):
+        new_data = {}
+        for name, val in data.items():
+            new_data[name] = _to_shared_memory(val)
+        return new_data
+    elif isinstance(data, list):
+        new_data = []
+        for val in data:
+            new_data.append(_to_shared_memory(val))
+        return new_data
+    elif isinstance(data, tuple):
+        new_data = []
+        for val in list(data):
+            new_data.append(_to_shared_memory(val))
+        return tuple(new_data)
+
+    # ignore other types
+    return data
+
+def _to_numpy_array(data):
+    """ Move all data objects back to numpy array
+
+    Parameters
+    ----------
+    data: dict/tuple/list of tensors
+        Data returned by user_parser
+    """
+    if th.is_tensor(data):
+        return data.numpy()
+    elif isinstance(data, np.ndarray):
+        return data # do nothing
+    elif isinstance(data, dict):
+        new_data = {}
+        for name, val in data.items():
+            new_data[name] = _to_numpy_array(val)
+        return new_data
+    elif isinstance(data, list):
+        new_data = []
+        for val in data:
+            new_data.append(_to_numpy_array(val))
+        return new_data
+    elif isinstance(data, tuple):
+        new_data = []
+        for val in list(data):
+            new_data.append(_to_numpy_array(val))
+        return tuple(new_data)
+
+    # ignore other types
+    return data
+
+def _estimate_sizeof(data):
+    """ Estimate the size of a data.
+        We assume the most memory consuming objects are tensors.
+
+    Parameters
+    ----------
+    data: dict/tuple/list of tensors
+        Data returned by user_parser
+    """
+    if th.is_tensor(data):
+        data_size = data.element_size() * data.nelement()
+    elif isinstance(data, np.ndarray):
+        data_size = data.size * data.itemsize
```

Comment:
if each element is an object, is this estimation still correct?

Reply:
No. But I think in most of the cases, it will be np arrays like features or node ids.